services/osparc-gateway-server/src/osparc_gateway_server/backend/osparc.py

```diff
@@ -1,298 +1,221 @@
-                self.log.info(
-                    "---> Service %s is started, and has ID %s",
-                    worker.name,
-                    service["ID"],
-                )
-                yield {"service_id": service["ID"]}
+            dashboard_address_arg_index = scheduler_cmd.index("--dashboard-address")
+            scheduler_cmd[dashboard_address_arg_index + 1] = "0.0.0.0:8787"
+        except ValueError:
+            scheduler_cmd.extend(["--dashboard-address", "0.0.0.0:8787"])
+        try:
+            scheduler_port_arg_index = scheduler_cmd.index("--port")
+            scheduler_cmd[scheduler_port_arg_index + 1] = "8786"
+        except ValueError:
+            scheduler_cmd.extend(["--port", "8786"])
+        self.log.debug("created scheduler command: %s", f"{scheduler_cmd=}")
+        # NOTE: the hostname of the gateway API must be modified so that the scheduler can
+        # send heartbeats to the gateway
+        scheduler_env.update(
+            {
+                "DASK_GATEWAY_API_URL": _replace_netloc_in_url(
+                    self.api_url, self.settings
+                ),
+            }
+        )
+        async for dask_scheduler_start_result in start_service(
+            self.docker_client,
+            self.settings,
+            self.log,
+            f"cluster_{cluster.id}_scheduler",
+            scheduler_env,
+            [c for c in self.cluster_secrets if c.cluster.name == cluster.name],
+            cmd=scheduler_cmd,
+            labels={"cluster_id": f"{cluster.id}"},
+        ):
+            yield dask_scheduler_start_result
+
+    async def do_stop_cluster(self, cluster: Cluster):
+        dask_scheduler_service_id = cluster.state.get("service_id")
+        await stop_service(self.docker_client, dask_scheduler_service_id, self.log)
+        await delete_secrets(self.docker_client, cluster)
+
+    async def do_check_clusters(self, clusters: List[Cluster]):
+        self.log.debug("--> checking clusters statuses: %s", f"{clusters=}")
+        ok = await asyncio.gather(
+            *[self._check_service_status(c) for c in clusters], return_exceptions=True
+        )
+        self.log.debug("<-- clusters status returned: %s", f"{ok=}")
+        return ok
 
-        except (DockerContainerError, DockerError):
-            self.log.exception(
-                "Unexpected Error while running container with parameters %s",
-                json.dumps(service_parameters, indent=2),
-            )
-            raise
-        except asyncio.CancelledError:
-            self.log.warn("Service creation was cancelled")
-            raise
+    async def do_start_worker(
+        self, worker: Worker
+    ) -> AsyncGenerator[Dict[str, Any], None]:
+        self.log.debug("received call to start worker as %s", f"{worker=}")
+        dask_scheduler_service_id = worker.cluster.state.get("service_id")
+        dask_scheduler = await self.docker_client.services.inspect(
```

Comment:
Is it safe to assume that this service is up and running at this point in time?

Reply:
well, I only had a quick look till now but I could not find a "dask-gateway" compatible way of raising an exception. I will have a second deeper look.

But it is so that when creating the cluster the API call only returns once the scheduler is up, so in theory is safe.

Reply:
ok used a dask-gateway-server PublicException in case of issue